Provide steps for Tesla Fleet API 2025 update.

andeart opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).

  Before submitting your pull request, please verify that you have chosen the correct target branch,
  and that the PR preview looks fine and does not include unrelated changes.
-->
## Proposed change
<!-- 
    Describe the big picture of your changes here to communicate to the
    maintainers why we should accept this pull request. If it fixes a bug
    or resolves a feature request, be sure to link to that issue in the 
    additional information section.
-->



## Type of change
<!--
    What types of changes does your PR introduce to our documentation/website?
    Put an `x` in the boxes that apply. You can also fill these out after
    creating the PR.
-->

- [ ] Spelling, grammar or other readability improvements (`current` branch).
- [X] Adjusted missing or incorrect information in the current documentation (`current` branch).
- [ ] Added documentation for a new integration I'm adding to Home Assistant (`next` branch).
  - [ ] I've opened up a PR to add logos and icons in [Brands repository](https://github.com/home-assistant/brands).
- [ ] Added documentation for a new feature I'm adding to Home Assistant (`next` branch).
- [ ] Removed stale or deprecated documentation.

## Additional information
<!--
    Details are important, and help maintainers processing your PR.
    Please be sure to fill out additional details, if applicable.
-->

- Link to parent pull request in the codebase: 
- Link to parent pull request in the Brands repository: 
- This PR fixes or closes issue: fixes [#134595](https://github.com/home-assistant/core/issues/134595) [#133961](https://github.com/home-assistant/core/issues/133961).

## Checklist
<!--
    Put an `x` in the boxes that apply. You can also fill these out after
    creating the PR. If you're unsure about any of them, don't hesitate to ask.
    We're here to help! This is simply a reminder of what we are going to look
    for before merging your code.
-->

- [X] This PR uses the correct branch, based on one of the following:
  - I made a change to the existing documentation and used the `current` branch.
  - I made a change that is related to an upcoming version of Home Assistant and used the `next` branch.
- [X] The documentation follows the Home Assistant documentation [standards].

[standards]: https://developers.home-assistant.io/docs/documenting/standards


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **Documentation**
  - Enhanced Tesla Fleet API integration instructions.
  - Added prerequisites for web domain hosting.
  - Expanded developer application setup guide with detailed steps.
  - Introduced a troubleshooting section for integration issues.
  - Updated documentation to reflect January 2025 API pricing changes.
  - Improved overall clarity and readability of the documentation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

andeart committed:
Add minor clarifications to Tesla Fleet integration

andeart committed:
Add detailed steps for Tesla authentication.

andeart committed:
Add clarity/formatting updates.

andeart committed:
Provide steps for 2025.1 update.

andeart committed:
Clarify integration error troubleshooting.

andeart committed:
Add clarifying step.

andeart committed:
Make changes recommended by coderabbit.

andeart committed:
Fix typo.

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

andeart commented on source/_integrations/tesla_fleet.markdown:
````diff
@@ -40,30 +40,124 @@
+   {"access_token":"LONG_GIBBERISH_ACCESS_TOKEN_STRING","expires_in":28800,"token_type":"Bearer"}%
+   ```
+   This is your access token. Copy everything between the double-quotes to be used next.
+4. Run the following CURL request:
````
This has been updated.

andeart commented on source/_integrations/tesla_fleet.markdown:
````diff
@@ -40,30 +40,124 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+   --data '{
+       "domain": "my.domain.com"
+   }'
+   ```
````
This has been updated.

rjgguk commented:
Should also include guidance around the use of EU URL's for non-US/Asia users.

Bre77 commented:
Some extra nuance, the billing step will not show up for all users, as Tesla isn't billing all countries yet.

Otherwise it looks great.

andeart commented:
@rjgguk Good catch. I'll add details for those specific parts.

@Bre77 That makes sense, I'll add a note mentioning that. Appreciate the quick review, folks!

andeart committed:
Add clarity around region-specific billing and API usage.

andeart committed:
Fix markdown linting issues.

andeart committed:
Merge branch 'clarify-tesla-fleet-integration-1' of github.com:andeart/home-assistant.io into clarify-tesla-fleet-integration-1

andeart committed:
Merge branch 'current' into clarify-tesla-fleet-integration-1

andeart commented on source/_integrations/tesla_fleet.markdown:
````diff
@@ -40,30 +40,135 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+
+2. Run this CURL request, replacing the variable values as specified in the notes below:
+  
+   ```shell
+   CLIENT_ID=REPLACE_THIS_WITH_YOUR_CLIENT_ID
+   CLIENT_SECRET=REPLACE_THIS_WITH_YOUR_CLIENT_SECRET
+   AUDIENCE="https://fleet-api.prd.na.vn.cloud.tesla.com"
+   curl --request POST \
+     --header 'Content-Type: application/x-www-form-urlencoded' \
+     --data-urlencode 'grant_type=client_credentials' \
+     --data-urlencode "client_id=$CLIENT_ID" \
+     --data-urlencode "client_secret=$CLIENT_SECRET" \
+     --data-urlencode 'scope=openid vehicle_device_data vehicle_cmds vehicle_charging_cmds' \
+     --data-urlencode "audience=$AUDIENCE" \
+     'https://fleet-auth.prd.vn.cloud.tesla.com/oauth2/v3/token'
+   ```
````
This suggestion is invalid, as the strings aren't guaranteed to follow any publicly-known pattern. And there is already a note below warning about escaping special characters.

andeart commented:
@Bre77 Details added, and ready for review. Let me know what you think please. Thanks!

Bre77 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,135 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
-When creating the application, you must set the redirect URL to `https://my.home-assistant.io/redirect/oauth`, but the other URLs can be set as desired. You must also complete both [step 3](https://developer.tesla.com/docs/fleet-api/getting-started/what-is-fleet-api#step-3-generate-a-public-private-key-pair) and [step 4](https://developer.tesla.com/docs/fleet-api/getting-started/what-is-fleet-api ...
+- A [Tesla](https://tesla.com) account
+- A [Developer Application](https://developer.tesla.com/en_US/dashboard)
+- A web domain that you own and can host your public key file directly onto without needing to rely on URL redirects. Simple file-hosting services such as AWS S3 may help with this for free/cheap.
```
I dont love mentioning _just_ AWS S3, we should either mention none or give a few different options. Core team might have specific ideas about this.

```suggestion
- A web domain and host that you can serve your public key file from. Some free options include:
  - [FleetKey.cc](https://fleetkey.cc)
  - [AWS S3](https://docs.aws.amazon.com/AmazonS3/latest/userguide/WebsiteHosting.html)
  - [Cloudflare Pages](https://pages.cloudflare.com/)
  - [Firebase Hosting](https://firebase.google.com/docs/hosting)
  
```

andeart replied:
That makes sense. I saw on the other thread that you built this- thanks for that! Applying your suggestion.

Bre77 commented on source/_integrations/tesla_fleet.markdown:
````diff
@@ -40,30 +40,135 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+
+2. Run this CURL request, replacing the variable values as specified in the notes below:
+  
+   ```shell
+   CLIENT_ID=REPLACE_THIS_WITH_YOUR_CLIENT_ID
+   CLIENT_SECRET=REPLACE_THIS_WITH_YOUR_CLIENT_SECRET
+   AUDIENCE="https://fleet-api.prd.na.vn.cloud.tesla.com"
+   curl --request POST \
+     --header 'Content-Type: application/x-www-form-urlencoded' \
+     --data-urlencode 'grant_type=client_credentials' \
+     --data-urlencode "client_id=$CLIENT_ID" \
+     --data-urlencode "client_secret=$CLIENT_SECRET" \
+     --data-urlencode 'scope=openid vehicle_device_data vehicle_cmds vehicle_charging_cmds' \
+     --data-urlencode "audience=$AUDIENCE" \
+     'https://fleet-auth.prd.vn.cloud.tesla.com/oauth2/v3/token'
+   ```
````
We need to cover EU here too, and the NA domain looks wrong.

````suggestion
   ```shell
   CLIENT_ID=REPLACE_THIS_WITH_YOUR_CLIENT_ID
   CLIENT_SECRET=REPLACE_THIS_WITH_YOUR_CLIENT_SECRET
   #AUDIENCE="https://fleet-api.prd.na.vn.cloud.tesla.com"
   #AUDIENCE="https://fleet-api.prd.eu.vn.cloud.tesla.com"
   #AUDIENCE="https://fleet-api.prd.cn.vn.cloud.tesla.com"
   curl --request POST \
     --header 'Content-Type: application/x-www-form-urlencoded' \
     --data-urlencode 'grant_type=client_credentials' \
     --data-urlencode "client_id=$CLIENT_ID" \
     --data-urlencode "client_secret=$CLIENT_SECRET" \
     --data-urlencode 'scope=openid vehicle_device_data vehicle_cmds vehicle_charging_cmds' \
     --data-urlencode "audience=$AUDIENCE" \
     'https://fleet-auth.prd.vn.cloud.tesla.com/oauth2/v3/token'
   ```
````

andeart replied:
@Bre77 Thanks for the suggestions.

1. I just verified the NA domain, and it looks correct. It appears unchanged in your suggestion too. Any idea what I may be missing?

2. There is already a line immediately below this section mentioning:
```
Replace the `AUDIENCE` value with your region-specific URL. The URL in the example is for users in North America and Asia-Pacific (excluding China). Refer to the [Base URLs documentation](https://developer.tesla.com/docs/fleet-api/getting-started/base-urls) for the URLs for other regions.
```
If we're instead expanding each region-specific curl command, shall I
    - also add the China URL in a similar but separate section? The Asia-Pacific URL excludes China.
    - abbreviate the line referencing the Tesla docs to something like `Refer to the [Base URLs documentation](https://developer.tesla.com/docs/fleet-api/getting-started/base-urls) for more info.`, omitting the direction to replace the `AUDIENCE` value?

Bre77 replied:
Oh I have made a mistake, the curl address is not region specific, I thought it was. Yes if you can condense this down to a single example that would be better, I just couldnt think of a great way to clearly communicate that.

andeart replied:
Sounds good. Does the single example that I've added suffice, along with the note after it to swap the `AUDIENCE` value to the correct URL (also provides links to where to find the region URLs)?

Bre77 commented on source/_integrations/tesla_fleet.markdown:
````diff
@@ -40,30 +40,135 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+3. The CURL request should return a response that looks something like:
+  
+   ```json
+   {"access_token":"LONG_GIBBERISH_ACCESS_TOKEN_STRING","expires_in":28800,"token_type":"Bearer"}
````
```suggestion
   {"access_token":"JSON_WEB_TOKEN","expires_in":28800,"token_type":"Bearer"}
```

Bre77 commented on source/_integrations/tesla_fleet.markdown:
````diff
@@ -40,30 +40,135 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+   ```shell
+   curl --location 'https://fleet-api.prd.na.vn.cloud.tesla.com/api/1/partner_accounts' \
+   --header 'Content-Type: application/json' \
+   --header 'Authorization: Bearer LONG_GIBBERISH_ACCESS_TOKEN_STRING' \
````
```suggestion
   --header 'Authorization: Bearer JSON_WEB_TOKEN' \
```

andeart committed:
Add hosting options.

Co-authored-by: Brett Adams <Bre77@users.noreply.github.com>

andeart committed:
Simplify token placeholder.

Co-authored-by: Brett Adams <Bre77@users.noreply.github.com>

andeart committed:
Simplify token placeholder.

Co-authored-by: Brett Adams <Bre77@users.noreply.github.com>

Bre77 reviewed (approved):
I am happy to get this in as is and we can take further input and feedback to refine it further.

andeart commented:
Thanks, @Bre77 ! I appreciate the assist on these changes. And let me know if I can help with testing any upcoming Tesla integration updates 🙂

parkerbxyz commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+- A web domain and host that you can serve your public key file from. Some free options include:
+  - [FleetKey.cc](https://fleetkey.cc)
+  - [AWS S3](https://docs.aws.amazon.com/AmazonS3/latest/userguide/WebsiteHosting.html)
+  - [Cloudflare Pages](https://pages.cloudflare.com/)
```
For what it's worth, I was able to use [Cloudflare R2](https://www.cloudflare.com/developer-platform/products/r2/) to host my public key, which was simpler to set up.

Bre77 replied:
This is the problem with listing options, there are so many that _could_ be there. I personally just want to push fleetkey.cc since that seems to be really helping people without the technical knowledge to do it themselves.

andeart commented:
@Bre77 My apologies for being squeaky. Thanks for approving the PR again. I see `At least 1 approving review is required by reviewers with write access` on this PR's thread.
I've read the [PR documentation](https://developers.home-assistant.io/docs/review-process/) which covers guidelines before a PR is reviewed, but I can't seem to find the next steps once a PR has been approved but waiting for a merge. Does your approval automatically send it to a queue for someone with write access to merge?

AGSPhoenix commented on source/_integrations/tesla_fleet.markdown:
````diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+   ```
+  
+   - If you had to change the `AUDIENCE` URL for your region in step 2, update the main domain of the `--location` arg.
+   - Replace `LONG_GIBBERISH_ACCESS_TOKEN_STRING` with the access token that you copied in the previous step.
````
The `LONG_GIBBERISH_ACCESS_TOKEN_STRING` placeholder seems to have been missed during editing.

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+3. Run this command to create and associate a public key with it: `openssl ec -in private-key.pem -pubout -out public-key.pem`
+4. Rename the public key file to `com.tesla.3p.public-key.pem`. This needs to be exactly this for Tesla's API to partner with your account correctly.
+5. Backup both these files somewhere safe and private for access later.
+6. Upload the public key file to your domain at the path `/.well-known/appspecific/com.tesla.3p.public-key.pem`. For example, if your domain is `https://my.domain.com`, the public key file must be accessible at `https://my.domain.com/.well-known/appspecific/com.tesla.3p.public-key.pem` . Do not use redirection logic to handle this, or the Tesla API will not recognize your app later in the process ...
```
```suggestion
6. Upload the public key file to your domain at the path `/.well-known/appspecific/com.tesla.3p.public-key.pem`. For example, if your domain is `https://my.domain.com`, the public key file must be accessible at `https://my.domain.com/.well-known/appspecific/com.tesla.3p.public-key.pem`. Do not use redirection logic to handle this, or the Tesla API will not recognize your app later in the process.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+
+These steps are also summarized in the [Tesla Fleet API documentation Step 2](https://developer.tesla.com/docs/fleet-api/getting-started/what-is-fleet-api#step-2-create-an-application), but it is recommended to follow the steps below to cover specific details.
+
+1. If you have not already, set up your [Tesla Developer account](https://developer.tesla.com/teslaaccount). Confirm that you have a verified email and multi-factor authentication set up.
```
```suggestion
1. If you have not already set up your [Tesla Developer account](https://developer.tesla.com/teslaaccount): Confirm that you have a verified email and multi-factor authentication set up.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+
+1. If you have not already, set up your [Tesla Developer account](https://developer.tesla.com/teslaaccount). Confirm that you have a verified email and multi-factor authentication set up.
+2. Navigate to the [Developer dashboard](https://developer.tesla.com/en_US/dashboard). Change your locale if needed to wherever your account is based, using the globe icon at the top-right corner.
+3. Click '**Create New Application**'. This should launch a new page with the header '**Create Fleet API Application**'.
```
```suggestion
3. Select **Create New Application**. This should launch a new page with the header **Create Fleet API Application**.
```
We favor [select over click](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/click)

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+1. If you have not already, set up your [Tesla Developer account](https://developer.tesla.com/teslaaccount). Confirm that you have a verified email and multi-factor authentication set up.
+2. Navigate to the [Developer dashboard](https://developer.tesla.com/en_US/dashboard). Change your locale if needed to wherever your account is based, using the globe icon at the top-right corner.
+3. Click '**Create New Application**'. This should launch a new page with the header '**Create Fleet API Application**'.
+4. For simple integrations, under '**Registration Type**', select '**Just for me**'. If you're confident about your situation being a business setup instead, select '**For my business**'. Click '**Next**'.
```
```suggestion
4. For simple integrations, under **Registration Type**, select **Just for me**. If you're confident about your situation being a business setup instead, select **For my business**. Select **Next**.
```
we use bold to mark UI strings. no quotes required.

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+2. Navigate to the [Developer dashboard](https://developer.tesla.com/en_US/dashboard). Change your locale if needed to wherever your account is based, using the globe icon at the top-right corner.
+3. Click '**Create New Application**'. This should launch a new page with the header '**Create Fleet API Application**'.
+4. For simple integrations, under '**Registration Type**', select '**Just for me**'. If you're confident about your situation being a business setup instead, select '**For my business**'. Click '**Next**'.
+5. At the '**Application Details**' step, use a name that is easy to refer to later, such as `ha-integration`. This will be needed later while configuring the integration. Click '**Next**'.
```
```suggestion
5. At the **Application Details** step, use a name that is easy to refer to later, such as `ha-integration`. This will be needed later while configuring the integration. Select **Next**.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+3. Click '**Create New Application**'. This should launch a new page with the header '**Create Fleet API Application**'.
+4. For simple integrations, under '**Registration Type**', select '**Just for me**'. If you're confident about your situation being a business setup instead, select '**For my business**'. Click '**Next**'.
+5. At the '**Application Details**' step, use a name that is easy to refer to later, such as `ha-integration`. This will be needed later while configuring the integration. Click '**Next**'.
+6. At the '**Client Details**' step, under '**Oauth Grant Type**', select '**Authorization Code and Machine-to-Machine**'.
```
```suggestion
6. At the **Client Details** step, under **Oauth Grant Type**, select **Authorization Code and Machine-to-Machine**.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+4. For simple integrations, under '**Registration Type**', select '**Just for me**'. If you're confident about your situation being a business setup instead, select '**For my business**'. Click '**Next**'.
+5. At the '**Application Details**' step, use a name that is easy to refer to later, such as `ha-integration`. This will be needed later while configuring the integration. Click '**Next**'.
+6. At the '**Client Details**' step, under '**Oauth Grant Type**', select '**Authorization Code and Machine-to-Machine**'.
+   - Under '**Allowed Origin URL(s)**', enter your domain where you hosted your public key earlier. Using the example above, this would be `https://my.domain.com/`.
```
```suggestion
   - Under **Allowed Origin URL(s)**, enter your domain where you hosted your public key earlier. Using the example above, this would be `https://my.domain.com/`.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+5. At the '**Application Details**' step, use a name that is easy to refer to later, such as `ha-integration`. This will be needed later while configuring the integration. Click '**Next**'.
+6. At the '**Client Details**' step, under '**Oauth Grant Type**', select '**Authorization Code and Machine-to-Machine**'.
+   - Under '**Allowed Origin URL(s)**', enter your domain where you hosted your public key earlier. Using the example above, this would be `https://my.domain.com/`.
+   - Under '**Allowed Redirect URI(s)**', enter `https://my.home-assistant.io/redirect/oauth`.
```
```suggestion
   - Under **Allowed Redirect URI(s)**, enter `https://my.home-assistant.io/redirect/oauth`.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+6. At the '**Client Details**' step, under '**Oauth Grant Type**', select '**Authorization Code and Machine-to-Machine**'.
+   - Under '**Allowed Origin URL(s)**', enter your domain where you hosted your public key earlier. Using the example above, this would be `https://my.domain.com/`.
+   - Under '**Allowed Redirect URI(s)**', enter `https://my.home-assistant.io/redirect/oauth`.
+   - Click '**Next**'.
```
```suggestion
   - Select **Next**.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+   - Under '**Allowed Origin URL(s)**', enter your domain where you hosted your public key earlier. Using the example above, this would be `https://my.domain.com/`.
+   - Under '**Allowed Redirect URI(s)**', enter `https://my.home-assistant.io/redirect/oauth`.
+   - Click '**Next**'.
+7. At the '**API & Scopes**' step, select the configurations you'd like to access.
```
```suggestion
7. At the **API & Scopes** step, select the configurations you'd like to access.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+   - Click '**Next**'.
+7. At the '**API & Scopes**' step, select the configurations you'd like to access.
+   - At least one of `Vehicle Information` or `Energy Product Information` **must** be selected for the integration to function. It is recommended you select all scopes for full functionality.
+   - If you're unsure, you can select only one of these two required scopes for now and update the scopes later from the Developer Dashboard. However, note that if the scopes are updated, you will need to reconfigure the integration fully (refer to the '**Integration is broken and needs to be reconfigured**' troubleshooting steps below).
```
```suggestion
   - If you're unsure, you can select only one of these two required scopes for now and update the scopes later from the Developer Dashboard. However, note that if the scopes are updated, you will need to reconfigure the integration fully (refer to the **Integration is broken and needs to be reconfigured** troubleshooting steps below).
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+7. At the '**API & Scopes**' step, select the configurations you'd like to access.
+   - At least one of `Vehicle Information` or `Energy Product Information` **must** be selected for the integration to function. It is recommended you select all scopes for full functionality.
+   - If you're unsure, you can select only one of these two required scopes for now and update the scopes later from the Developer Dashboard. However, note that if the scopes are updated, you will need to reconfigure the integration fully (refer to the '**Integration is broken and needs to be reconfigured**' troubleshooting steps below).
+   - Click '**Next**'.
```
```suggestion
   - Select **Next**.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+   - At least one of `Vehicle Information` or `Energy Product Information` **must** be selected for the integration to function. It is recommended you select all scopes for full functionality.
+   - If you're unsure, you can select only one of these two required scopes for now and update the scopes later from the Developer Dashboard. However, note that if the scopes are updated, you will need to reconfigure the integration fully (refer to the '**Integration is broken and needs to be reconfigured**' troubleshooting steps below).
+   - Click '**Next**'.
+8. At the '**Billing Details (Optional)**' step, enter your billing details if needed, then click '**Submit**'. Tesla does not support billing in all countries yet. **Developers in countries that do not yet support payments will not see this step**.
```
```suggestion
8. At the **Billing Details (Optional)** step, enter your billing details if needed, then select **Submit**. Tesla does not support billing in all countries yet. **Developers in countries that do not yet support payments will not see this step**.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+   - Click '**Next**'.
+8. At the '**Billing Details (Optional)**' step, enter your billing details if needed, then click '**Submit**'. Tesla does not support billing in all countries yet. **Developers in countries that do not yet support payments will not see this step**.
+   - Tesla provides a $10 monthly credit for personal API usage, so your level of usage may be covered for free. Detailed pricing info for commands, data polling, and wake signals can be found at [developer.tesla.com](https://developer.tesla.com). Use these details to calculate your usage estimate if you rely heavily on this integration.
+   - If unsure, you can click '**Skip & Submit**' for now and add the billing details later when your usage is close to the free threshold.
```
```suggestion
   - If unsure, you can select **Skip & Submit** for now and add the billing details later when your usage is close to the free threshold.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+The following steps involve sensitive credentials. Never share your `Client Secret` or access token with anyone directly, and ensure you're working in a secure environment.
+{% endwarning %}
+
+1. Get your OAuth details by going to your [Developer dashboard](https://developer.tesla.com/en_US/dashboard), clicking '**View Details**' under the app you set up for Home Assistant integration, then clicking on the '**Credentials & APIs**' tab. Note the `Client ID` and `Client Secret` strings.
```
```suggestion
1. Get your OAuth details by going to your [Developer dashboard](https://developer.tesla.com/en_US/dashboard). Under the app you set up for Home Assistant integration select **View Details**. Then, select the **Credentials & APIs** tab. Note the `Client ID` and `Client Secret` strings.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+
+{% details "Linking the Developer Application with Home Assistant" %}
+
+1. Get your OAuth details by going to your [Developer dashboard](https://developer.tesla.com/en_US/dashboard), clicking '**View Details**' under the app you set up for Home Assistant integration, then clicking on the '**Credentials & APIs**' tab. Note the `Client ID` and `Client Secret` strings, these will be needed later.
```
```suggestion
1. Get your OAuth details by going to your [Developer dashboard](https://developer.tesla.com/en_US/dashboard). Select **View Details** under the app you set up for Home Assistant integration. Then, select the **Credentials & APIs** tab. Note the `Client ID` and `Client Secret` strings, these will be needed later.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+{% details "Linking the Developer Application with Home Assistant" %}
+
+1. Get your OAuth details by going to your [Developer dashboard](https://developer.tesla.com/en_US/dashboard), clicking '**View Details**' under the app you set up for Home Assistant integration, then clicking on the '**Credentials & APIs**' tab. Note the `Client ID` and `Client Secret` strings, these will be needed later.
+2. In Home Assistant, the integration wizard should walk you through the default steps. If not already started, scroll above and click the '**ADD INTEGRATION TO MY**' button to start the integration wizard. The integration will ask you for all of the necessary integration configuration.
```
```suggestion
2. In Home Assistant, the integration wizard should walk you through the default steps. If not already started, scroll above and select the **ADD INTEGRATION TO MY** button to start the integration wizard. The integration will ask you for all of the necessary integration configurations.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+
+1. Get your OAuth details by going to your [Developer dashboard](https://developer.tesla.com/en_US/dashboard), clicking '**View Details**' under the app you set up for Home Assistant integration, then clicking on the '**Credentials & APIs**' tab. Note the `Client ID` and `Client Secret` strings, these will be needed later.
+2. In Home Assistant, the integration wizard should walk you through the default steps. If not already started, scroll above and click the '**ADD INTEGRATION TO MY**' button to start the integration wizard. The integration will ask you for all of the necessary integration configuration.
+3. In the '**Add credentials**' step in the wizard, enter your Tesla Fleet developer application name (from step 5 in the '**Setting up the Developer Application**' section above), and the Oauth Client ID and Client Secret (from step 1 above). Click '**Submit**'.
```
```suggestion
3. In the **Add credentials** step in the wizard, enter your Tesla Fleet developer application name (from step 5 in the **Setting up the Developer Application** section above), and the Oauth Client ID and Client Secret (from step 1 above). Select **Submit**.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+2. In Home Assistant, the integration wizard should walk you through the default steps. If not already started, scroll above and click the '**ADD INTEGRATION TO MY**' button to start the integration wizard. The integration will ask you for all of the necessary integration configuration.
+3. In the '**Add credentials**' step in the wizard, enter your Tesla Fleet developer application name (from step 5 in the '**Setting up the Developer Application**' section above), and the Oauth Client ID and Client Secret (from step 1 above). Click '**Submit**'.
+4. At this step, you should be taken to the Tesla authentication page. You will need to re-enter your Tesla account login credentials.
+5. At the confirmation page with the header '**Allow ha-integration access to your Tesla Account**' (the specific application name will be whatever you set earlier), click the '**Select All**' button. This list of scopes is already limited to the specific scopes you chose for the application information earlier, so it is not necessary to review them again. Click '**Allow**'.
```
```suggestion
5. At the confirmation page with the header **Allow ha-integration access to your Tesla Account** (the specific application name will be whatever you set earlier), select the **Select All** button. This list of scopes is already limited to the specific scopes you chose for the application information earlier, so it is not necessary to review them again. Select **Allow**.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
+3. In the '**Add credentials**' step in the wizard, enter your Tesla Fleet developer application name (from step 5 in the '**Setting up the Developer Application**' section above), and the Oauth Client ID and Client Secret (from step 1 above). Click '**Submit**'.
+4. At this step, you should be taken to the Tesla authentication page. You will need to re-enter your Tesla account login credentials.
+5. At the confirmation page with the header '**Allow ha-integration access to your Tesla Account**' (the specific application name will be whatever you set earlier), click the '**Select All**' button. This list of scopes is already limited to the specific scopes you chose for the application information earlier, so it is not necessary to review them again. Click '**Allow**'.
+6. You should now see a Home Assistant page asking if you would like to '**Link account to Home Assistant?**'. Click '**Link account**'.
```
```suggestion
6. You should now see a Home Assistant page asking if you would like to **Link account to Home Assistant?**. Select **Link account**.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -40,30 +40,140 @@ The Tesla Fleet API {% term integration %} exposes various sensors from Tesla ve
-For more details please see [developer.tesla.com](https://developer.tesla.com).
+For more details, please see [developer.tesla.com](https://developer.tesla.com).
+
+Note that Tesla does not support billing in all countries yet. **Developers in countries that do not yet support payments will not be able to review their billing or usage**. For countries that do support billing, the current billing usage can be viewed at any time by going to your [Developer Dashboard](https://developer.tesla.com/en_US/dashboard), clicking '**View Details**' under the app you se ...
```
```suggestion
Note that Tesla does not support billing in all countries yet. **Developers in countries that do not yet support payments will not be able to review their billing or usage**. For countries that do support billing, the current billing usage can be viewed at any time by going to your [Developer Dashboard](https://developer.tesla.com/en_US/dashboard), select **View Details** under the app you set up for Home Assistant integration. Then, select the **Application Usage** tab.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -232,3 +342,22 @@ The Tesla Fleet API only provides power data for Powerwall and Solar products. T
+  2. For some errors in the integration, waking the vehicle or energy product through the official app, then restarting Home Assistant will sometimes help re-authenticate the connection with Tesla's API. You may be asked to re-authenticate through the Tesla website.
+
+- **Integration is broken or needs to be reconfigured**
+  1. Ensure that you have a Tesla developer application ready for usage (refer to the instructions in the '**Setting up the Developer Application**' section above).
```
```suggestion
  1. Ensure that you have a Tesla developer application ready for usage (refer to the instructions in the **Setting up the Developer Application** section above).
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -232,3 +342,22 @@ The Tesla Fleet API only provides power data for Powerwall and Solar products. T
+
+- **Integration is broken or needs to be reconfigured**
+  1. Ensure that you have a Tesla developer application ready for usage (refer to the instructions in the '**Setting up the Developer Application**' section above).
+  2. Go to your Tesla Fleet integration page in Home Assistant, then click the '**RECONFIGURE**' button to bring the integration wizard up.
```
```suggestion
  2. Go to your Tesla Fleet integration page in Home Assistant, then select the **Reconfigure** button to bring the integration wizard up.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -232,3 +342,22 @@ The Tesla Fleet API only provides power data for Powerwall and Solar products. T
+- **Integration is broken or needs to be reconfigured**
+  1. Ensure that you have a Tesla developer application ready for usage (refer to the instructions in the '**Setting up the Developer Application**' section above).
+  2. Go to your Tesla Fleet integration page in Home Assistant, then click the '**RECONFIGURE**' button to bring the integration wizard up.
+     - If the '**RECONFIGURE**' button is not visible, clear any Application Credentials related to Tesla Fleet from your Application Credentials page (can be found at `http://homeassistant.local:PORT/config/application_credentials`), then restart Home Assistant. After the restart, navigate to the Tesla Fleet integration page and the '**RECONFIGURE**' button should be visible.
```
```suggestion
     - If the **Reconfigure** button is not visible, clear any Application Credentials related to Tesla Fleet from your Application Credentials page (can be found at `http://homeassistant.local:PORT/config/application_credentials`), then restart Home Assistant. After the restart, navigate to the Tesla Fleet integration page, and the **Reconfigure ** button should be visible.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -232,3 +342,22 @@ The Tesla Fleet API only provides power data for Powerwall and Solar products. T
+  1. Ensure that you have a Tesla developer application ready for usage (refer to the instructions in the '**Setting up the Developer Application**' section above).
+  2. Go to your Tesla Fleet integration page in Home Assistant, then click the '**RECONFIGURE**' button to bring the integration wizard up.
+     - If the '**RECONFIGURE**' button is not visible, clear any Application Credentials related to Tesla Fleet from your Application Credentials page (can be found at `http://homeassistant.local:PORT/config/application_credentials`), then restart Home Assistant. After the restart, navigate to the Tesla Fleet integration page and the '**RECONFIGURE**' button should be visible.
+  3. Follow the steps in the '**Linking the Developer Application with Home Assistant**' section above.
```
```suggestion
  3. Follow the steps in the **Linking the Developer Application with Home Assistant** section above.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -232,3 +342,22 @@ The Tesla Fleet API only provides power data for Powerwall and Solar products. T
+  3. Follow the steps in the '**Linking the Developer Application with Home Assistant**' section above.
+
+- **Integration no longer works after the January 2025 API pricing updates**
+  1. Refer to the '**Integration is broken**' troubleshooting steps above.
```
```suggestion
  1. Refer to the **Integration is broken** troubleshooting steps above.
```

c0ffeeca7 commented on source/_integrations/tesla_fleet.markdown:
```diff
@@ -232,3 +342,22 @@ The Tesla Fleet API only provides power data for Powerwall and Solar products. T
+  1. Refer to the '**Integration is broken**' troubleshooting steps above.
+
+- **Integration shows `a condition has not been met to process the request`**
+  1. Confirm that you've run all the steps from both the '**Hosting a Public/Private Key Pair**' and '**Register your application as a Fleet API partner**' sections above.
```
```suggestion
  1. Confirm that you've run all the steps from both the **Hosting a Public/Private Key Pair** and **Register your application as a Fleet API partner** sections above.
```

c0ffeeca7 committed:
Tiny style tweaks

2 modifications:
- we use bold to mark UI strings, no quotes required
- we [favor select over click](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/click)

c0ffeeca7 reviewed (approved):
Thank you, @andeart and @Bre77 :+1:

Bre77 commented:
FYI @andeart I'm also making command signing better by removing the Bluetooth process. https://github.com/home-assistant/home-assistant.io/pull/36775
